duckduckgo/macos-browser
# Add session restore setting and menu item

Task/Issue URL: https://app.asana.com/0/72649045549333/1202238581985189/f
Tech Design URL: https://app.asana.com/0/481882893211075/1202279173511424/f 
Tech Design URL: https://app.asana.com/0/481882893211075/1202290603284019/f
CC: @mallexxx @samsymons 

**Description**:
Functional changes:
* Settings screen was updated with a new setting to control automatic session restoration. It's disabled by default (a change from current behaviour). Also wording was updated in the old "Default Browser" pane, now called "Startup".
* _Reopen All Windows from Last Session_ menu item was added:
  * it is only enabled when the user did not turn on automatic session restoring, and when they have a session to restore
  * the ⌘⇧T shortcut is moved to this item from _Reopen Last Closed Tab_ when the former is active immediately,
    after the app has started. As soon as user starts interacting with the browser (opening tabs, navigating to websites),
    the shortcut is moved back to _Reopen Last Closed Tab_.
* The session is forcibly restored from a persisted state when the app is being automatically updated via Sparkle
* The persisted previous session is kept encrypted in app memory for the lifetime of the current session, unless any of
  the following happens in which case the previous session is cleared:
  * the user restores it,
  * the user uses fire button in any situation.

Technical changes:
* I updated fire animation view to load its animation (1.3MB JSON file) on a background thread, as this view used to be
  loaded together with the window, contributing to the startup time of the app
* I ended up implementing Sparkle restart handling using `updaterWillRelaunchApplication` (instead of didRelaunch, as agreed in the Tech Design). This is because didRelaunch callback is called only after `applicationDidFinishLaunching` so to handle it that way we'd have to perform startup tasks after a delay which could have caused various types of weird issues going forward. [Sparkle handle it via UserDefaults as well](https://github.com/sparkle-project/Sparkle/pull/1115/files), but I added our own flag for this, to not be dependent on their private implementation.
* `FileStore` protocol got a new method called `decrypt`, and `loadData` now takes additional `decryptIfNeeded` parameter, defaulting to true (to comply with existing use cases). This flag is set to false when loading persisted state into memory.

Changes to scripts:
* I added `sparkle-sandbox.sh` to simplify testing automatic updates via Sparkle. Please consult readme for more details on its features and usage
* I updated archive script (and Sparkle script) to always execute from a temporary directory (having previously been automatically copied to a new temporary directory). This allows for modifying the scripts or switching branches while the script is running.
* I changed output DMG image name to lowercase (`duckduckgo-x.y.z.dmg`)



**Steps to test this PR**:
1. Clear app data using `./clean-app.sh debug`
1. Run the app, open some tabs/windows
1. Restart the app, verify that it starts with a single window with new tab page
1. Press ⌘⇧T and verify that the app opens windows from previous session
1. Press ⌘⇧T again and verify that nothing happens
1. Close a tab, press ⌘⇧T and verify that it reopens a closed tab
1. Restart the app and open (another) new tab, press ⌘⇧T and verify that nothing happens (state is not restored because the user started interacting with the app)
1. Use fire button to burn a single tab, single window and all data; verify that in all cases you can't restore previous session
1. Go to Settings -> Startup and enable automatic session restoring
1. Restart the app and verify that the previous session is automatically restored; verify that you can't re-restore the session using the menu item
1. Test updating with Sparkle. See `scripts/README.md`, set up ngrok and Apple app-specific password, run the script and follow instructions. Open the "old" version of the app, make sure automatic session restore is disabled, create a session that can be restored, check for updates, install the new app and verify that the session is restored on the app restart.

**Testing checklist**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**
